src/manifold/deferred.clj

```diff
@@ -1115,10 +1115,16 @@
             (success-error-unrealized x
               val (success! d val)
               err (error! d err)
-              (do (on-realized (chain' x)
-                    #(success! d %)
-                    #(error! d %))
-                  (recur (inc i))))
+              (let [l (listener #(success! d %)
+                                #(error! d %))]
+                (add-listener! x l)
+                (on-realized d
+                             (fn [&args]
+                               (cancel-listener! x l))
+                             (fn [&args]
+                               (cancel-listener! x l)))
```

Comment:
@solatis I would avoid doing `[&args]` here, since that's actually a valid symbol name. This will break if more than one arg is passed. Try something like `[& _]` instead.

Reply:
Also, if I'm not mistaken, a listener can be shared, so you can move its creation outside the loop and reuse it for each alternative.